[Index Management] Improve error messages when editing index settings

sabarasaba opened the pull request:
Fixes https://github.com/elastic/kibana/issues/118026

### Summary
Currently when there's an error while saving index settings a very non informative error is shown to the user below the JSON editor. Which makes it harder for users to find if the viewport of their screen is narrow. This PR improves this error message by wrapping it in an error callout above the JSON editor and showing the actual error message returned by Elasticsearch instead of just a `Bad Request` label.

**How to test**
1. Start elasticsearch with `yarn es snapshot --license=trial` 
2. Start kibana with `yarn start`
3. Navigate to `Stack Management` -> `Index Management` -> open the flyout of any index and go to the `Edit settings` tab
4. Set `index.routing.allocation.include._tier_preference` to `"[data_content]"`
4. Verify that clicking the save button will show an error

**Screenshots**

<details>
<summary>before/after</summary>

// Old version
![Screenshot 2021-11-19 at 13 38 34](https://user-images.githubusercontent.com/1191206/142624128-3d31cfb3-bcd1-42ea-b720-fa43fb8327c9.png)

// Improved error handling
![Screenshot 2021-11-19 at 13 06 27](https://user-images.githubusercontent.com/1191206/142624484-38f61d86-3122-4174-84df-2e4f6731a4f4.png)
![Screenshot 2021-11-19 at 13 06 17](https://user-images.githubusercontent.com/1191206/142624478-119a5956-52df-4a80-be1b-4b9f34103d54.png)
</details>

sabarasaba commented on x-pack/plugins/index_management/__jest__/client_integration/home/indices_tab.test.ts:
```diff
@@ -12,6 +12,33 @@ import { setupEnvironment, nextTick } from '../helpers';
+// component but an editor being instantiated on a div reference, we cannot mock
+// the component and replace it with something else. In this particular case we're
+// mocking the returned instance of the editor to always have the same values.
+jest.mock('../../../public/application/lib/ace.js', () => {
```
I couldn't figure out what was the best way of testing this, the way the editor is instantiated is through [a div ref](https://github.com/elastic/kibana/blob/c72eb357650de4a90399420e015c9d905df3738a/x-pack/plugins/index_management/public/application/lib/ace.js#L47) which makes it hard to mock other than creating a ghost instance such as this one I implemented.

sebelga replied:
Makes sense to me. Until we have a dedicated component we can mock this is the best we can do IMO.

What do you think about having the option to set the returned value from the ace editor from inside the test. It would make it clearer why in this case we return an invalid tier preference.

Something like this

```js
const mockGetAceEditorValue = jest.fn().mockReturnValue(`{}`);

jest.mock('../../../public/application/lib/ace.js', () => {
  const createAceEditor = () => {
    return {
      getValue: mockGetAceEditorValue,
      getSession: () => {
     ...

// And then in the specific test

test('shows error callout when request fails', async () => {
  const { actions, find, component, exists } = testBed;

  mockGetAceEditorValue.mockReturnValue(`{
    "index.routing.allocation.include._tier_preference": "non_existent_tier"
  }`);

...
```

sabarasaba commented:
@elasticmachine merge upstream

elasticmachine commented:
Pinging @elastic/kibana-stack-management (Team:Stack Management)

sabarasaba commented:
@elasticmachine merge upstream

sabarasaba commented:
@elasticmachine merge upstream

sebelga commented on x-pack/plugins/index_management/public/application/services/api.ts:
```diff
@@ -194,15 +194,15 @@ export async function loadIndexSettings(indexName: string) {
+    body: JSON.stringify(body),
+  }).then((result) => {
+    // Only track successful requests.
+    uiMetricService.trackMetric(METRIC_TYPE.COUNT, UIM_UPDATE_SETTINGS);
```
Looking at `sendRequest` it wraps the HTTP request behind a `try / catch`. This means that this Promise will always succeed. We should then wrap the `uiMetricService.trackMetric` call inside an

```js
if (result.error === null) { ... }
```

sabarasaba replied:
Totally right!

sebelga reviewed (approved):
Tested locally and works great. Great improvement @sabarasaba ! Thanks for enhancing this.
I left a few non blocker comments.

sabarasaba commented:
Thanks for having a look @sebelga! I've addressed your points with https://github.com/elastic/kibana/pull/119144/commits/848f740507eb3079c66a4712696cbe85a4f0cc88

sebelga reviewed (approved):
LGTM! Thanks for making the changes 👍

cuff-links commented:
@sabarasaba Howdy. I think we need this to be backported to to 7.17 since all of the other tests being backported include changes made in this PR.
